Refactor and rename confusing pthreads async functions

kripken opened the pull request:
As noted in #9086, `emscripten_async_queue_on_thread` is a
misleading name since it is only async if calling another thread -
if we are already on the same thread, we just do the call
immediately.

Fixing that by making it actually async isn't an option, since

1. If the current thread never returns to the main event loop,
   an async event would never happen. We have to fire it
   synchronously in that case, but we don't know if that is the
   case or not (it depends on user code).
1. It would be less efficient for the common case.

This PR renames `emscripten_async_queue_on_thread` to
`emscripten_call_on_thread_maybe_async`. The `maybe_async`
part indicates that it may be async but won't always be (again,
it is async if we are going to run on another thread). For
completeness, a new `emscripten_call_on_thread_async` is
added which is always asynchronous, even on the main thread.
That can be a simpler model to reason about for users.

That is a breaking change for people using that API. I would
guess not many have since while it is in `threading.h` it isn't
documented elsewhere. And this PR should help some of those
users (including the one that suggested this PR), so overall
it seems worth a breaking change.



Fixes #9086

dschuff commented:
Maybe these functions should have a return value indicating whether they operated synchronously or not? This looks similar to many of the Pepper APIs; most of these can return an error value (either as a return from the call or as an argument to the completion callback) and one of the possible error values is that the operation is happening async. Basically every use of these APIs needs to check whether the operation is actually done or not, and I imagine the same would be true of these APIs.

juj commented on system/include/emscripten/threading.h:
```diff
@@ -332,10 +332,17 @@ EMSCRIPTEN_RESULT emscripten_wait_for_call_i(em_queued_call *call, double timeou
 
-#define emscripten_async_queue_on_thread(target_thread, sig, func_ptr, satellite, ...) emscripten_async_queue_on_thread_((target_thread), (sig), (void*)(func_ptr), (satellite),##__VA_ARGS__)
+// Queues the given function call to be performed on the specified thread. If
+// it is already on the right thread, it will just execute it; otherwise it
```
"If it is already on the right thread" reads a bit vague. Perhaps instead `"If the target thread is the same as the calling thread, the function will be immediately executed before returning."?

kripken replied:
Thanks, yeah, that was unclear... fixing.

juj commented:
> This PR renames `emscripten_async_queue_on_thread` to
> `emscripten_call_on_thread_maybe_async`.

Somewhat slight dislike towards words like `maybe` in general, it is a bit undescriptive. Also, `maybe async` has a ring to it meaning "most of the time sync, but maybe sometimes async", which is not the intent of the function - it is always async if dispatching to another thread, but always sync if dispatching to the calling thread.

If this is something that documenting it clearer doesn't fix, perhaps a function name like `emscripten_queue_on_thread` or `emscripten_dispatch_to_thread` might work to avoid the problem? Or if neither of those sound good and someone wants that we have a strictly accurate name, let's go with `emscripten_call_on_thread_async_to_caller_sync`?

kripken commented:
@dschuff 

> Maybe these functions should have a return value indicating whether they operated synchronously or not? 

Good idea! Added a return value now, 1 meaning "executed", 0 meaning "will execute asynchronously later on that thread".

@juj 

> perhaps a function name like emscripten_queue_on_thread or emscripten_dispatch_to_thread

Oh nice, that's a much better idea! Changed to `emscripten_dispatch_to_thread` now.

kripken commented on system/lib/pthread/library_pthread.c:
```diff
@@ -846,13 +848,12 @@ em_queued_call* emscripten_async_waitable_run_in_main_runtime_thread_(
   em_queued_call* q = em_queued_call_malloc();
   assert(q);
-  if (!q)
-    return;
```
I removed this as it seems incorrect - we have an assertion right above it? With this return we would just silently drop messages when out of memory.

juj replied:
Hmm, this does not look good to remove. Asserts will not be present in release builds. The issue here is that if malloc fails, we will be unable to queue the message. When we are in such a constrained memory situation, the only thing we can do is to silently cancel, hoping that whatever code path we have elsewhere that saw the malloc failure, would be able to deal with the OOM situation.

If we do want to change this to something else, it should be something along the lines of
```
if (!q) abortCurrentThread();
```
The original behavior of `if (!q) return;` would cause the calling thread to hang indefinitely, waiting for a message to complete that never will, which is a relatively good way to park the calling thread to a "we have hung now" state.

kripken replied:
Ok, I restored the old code, and added a TODO.

This pattern appears in a bunch of places in this code and might be worth changing. Seems like for async messages it just drops them silently, not even hanging, atm?

juj replied:
Yeah, async messages would be silently dropped. Given that the page is in a tight OOM state, it's likely that something else will soon OOM anyways (if it hasn't already), which is why I did not worry about that too much. But adding an explicit abort here would be ok.

kripken commented:
ping @juj, @dschuff - I believe all review comments have been addressed?

dschuff commented on system/include/emscripten/threading.h:
```diff
@@ -332,10 +332,19 @@ EMSCRIPTEN_RESULT emscripten_wait_for_call_i(em_queued_call *call, double timeou
+// otherwise.
+#define emscripten_dispatch_to_thread(target_thread, sig, func_ptr, satellite, ...) _emscripten_call_on_thread(0, (target_thread), (sig), (void*)(func_ptr), (satellite),##__VA_ARGS__)
+
+// Similar to emscripten_queue_on_thread_maybe_async, but always runs the
```
```suggestion
// Similar to emscripten_dispatch_to_thread, but always runs the
```

dschuff commented on system/lib/pthread/library_pthread.c:
```diff
@@ -883,7 +891,18 @@ void EMSCRIPTEN_KEEPALIVE emscripten_async_queue_on_thread_(
   // been performed.
   q->calleeDelete = 1;
-  emscripten_async_queue_call_on_thread(targetThread, q);
+  // The maybe_async() function will not be async on the same thread; force
```
does `maybe_async` need to be updated here too?

kripken replied:
Thanks, yes, updating.

dschuff commented on tests/pthread/call_async.c:
```diff
@@ -0,0 +1,35 @@
+
+int main() {
+  assert(state == 0);
+  // This maybe_async call will be synchronous since we are on the right
```
```suggestion
  // This dispatch_to_thread call will be synchronous since we are on the right
```
